### PR TITLE
[Feature] Add MultiImageMixDataset

### PR DESCRIPTION
**Pull request by lkm2835**
## Modification

- `mmseg/datasets/builder.py`: Add (if cfg['type'] == 'MultiImageMixDataset'])

- `mmseg/datasets/dataset_wrappers.py`: Add class MultiImageMixDataset

-  `tests/test_data/test_dataset.py`: Add unittests

## Use cases (Optional)

```
train_pipeline = [
    dict(type='Mosaic'),
    dict(type='Resize', img_scale=(1024, 512), keep_ratio=True),
    dict(type='RandomFlip', prob=0.5),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_semantic_seg']),
]

train_dataset = dict(
    type='MultiImageMixDataset',
    dataset=dict(
        classes=classes,
        palette=palette,
        type=dataset_type,
        reduce_zero_label=False, 
        img_dir=data_root + "images/train",
        ann_dir=data_root + "annotations/train",
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(type='LoadAnnotations'),
        ]  
    ),
    pipeline=train_pipeline
)
```

[use case](https://github.com/open-mmlab/mmdetection/blob/f08548bfd6d394a82566022709b5ce9e6b0a855e/configs/yolox/yolox_s_8x8_300e_coco.py) in mmdet

----
Original code: [MultiImageMixDataset](https://github.com/open-mmlab/mmdetection/blob/f08548bfd6d394a82566022709b5ce9e6b0a855e/mmdet/datasets/dataset_wrappers.py#L289) in mmdet
Related: [Issue#1045](https://github.com/open-mmlab/mmsegmentation/issues/1045), [Pull Request#1093](https://github.com/open-mmlab/mmsegmentation/pull/1093)

### CODE REVIEW

**Comment by MengzhangLI** [comment]
Hi, @lkm2835 thanks for your pr.

We would review it ASAP. 

Best,

**Comment by Younghoon-Lee** [comment]
image

This is the sample image that is built by `MultiImageMixDataset` from our custom dataset. 
As you can see, four images come together to form a single square.

**Comment by Junjun2016** [comment]
> image
> 
> This is the sample image that is built by `MultiImageMixDataset` from our custom dataset. As you can see, four images come together to form a single square.

Could you please provide some results on your custom dataset?

**Comment by Younghoon-Lee** [comment]
> Could you please provide some results on your custom dataset?

Okay I see . We will do some tests with and without Mosaic. Please note that our custom dataset has a few noises in annotation.

File: mmseg/datasets/dataset_wrappers.py [hunk 1]
```diff
@@ -188,3 +190,87 @@ def __getitem__(self, idx):
...
+                transform = build_from_cfg(transform, PIPELINES)
+                self.pipeline.append(transform)
+            else:
+                raise TypeError('pipeline must be a dict')
+
+        self.dataset = dataset
+        self.CLASSES = dataset.CLASSES
+        self.PALETTE = dataset.PALETTE
+        if hasattr(self.dataset, 'flag'):
+            self.flag = dataset.flag
```
Comment (Junjun2016) [inline mmseg/datasets/dataset_wrappers.py:235]: There is no `flag` attribute in mmseg dataset

File: tools/browse_dataset.py [hunk 2]
```diff
@@ -127,16 +127,17 @@ def retrieve_data_cfg(config_path, skip_type, show_origin=False):
     train_data_cfg = cfg.data.train
```
Comment (Junjun2016) [inline tools/browse_dataset.py:127]: Can also add an argument named `options`, refer to https://github.com/open-mmlab/mmdetection/blob/f08548bfd6/tools/misc/browse_dataset.py#L36 and https://github.com/open-mmlab/mmsegmentation/blob/master/tools/train.py#L71.

**Comment by Junjun2016** [comment]
> > image
> > This is the sample image that is built by `MultiImageMixDataset` from our custom dataset. As you can see, four images come together to form a single square.
> 
> Could you please provide some results on your custom dataset?

It seems that you have set the opacity to 0, can set it to 0.5 to also show the label annotations.

**Comment by Junjun2016** [comment]
Please improve the unittests coverage.

**Comment by Junjun2016** [comment]
Hi @RockeyCoss
Please review it and do some ablation studies.

**Comment by Younghoon-Lee** [comment]
We made changes you requested. And I browsed about 100 images and annotations and they seemed all fine :)

**Comment by Junjun2016** [comment]
> We made changes you requested. And I browsed about 100 images and annotations and they seemed all fine :)

Thanks for your hard work.
We could do some ablation studies together.
Looking forward to the result of your project.

**Comment by RockeyCoss** [comment]
Thanks for your contribution. I will combine it with Mosaic data augmentation and do some experiments.

**Comment by lkm2835** [comment]
`mmseg/datasets/dataset_wrappers.py` line 247 ~ 259

When it merged with Mosaic, it would be easy to improve the unittests coverage.

**Comment by Junjun2016** [comment]
> > Thanks for your hard work. We could do some ablation studies together. Looking forward to the result of your project.
> 
> Sorry for taking so late. I brought some results below. All results were applied of 512 img scale for Mosaic augmentation.
> 
> Mosaic_prob	best	best epoch	last	last epoch
> 0	0.5752	25	0.5685	50
> 0.5	0.5728	35	0.5696	50
> 1.0	0.5814	49	0.579	50
> We have done different img scales test for Mosaic augmentation as well, but 512x512 showed best performance. (Mosaic prob =1.0)
> 
> img_scale	best	best epoch	last	last epoch
> 384	0.5297	38	0.5269	50
> 448	0.5632	49	0.5627	50
> 512	0.5814	49	0.579	50
> 576	0.5475	41	0.5405	50
> 640	0.544	46	0.5411	50
> According to test results, Mosaic augmentation seems to show better performance on a few label images.

What's the best result on your custom dataset (SOTA)?

**Comment by Younghoon-Lee** [comment]
> What's the best result on your custom dataset (SOTA)?

The best result is 0.5814 ,which is 512x512 img scale for Mosaic(prob=1) augmentation. 

Here is config script for more detail. 

```
train_pipeline = [
    dict(type='RandomMosaic', prob=1, img_scale=(512,512)),
    dict(type='RandomCrop', crop_size=(512, 512)),
    dict(type='RandomFlip', prob=0.5),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_semantic_seg']),
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(512,512),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(type='Normalize', **img_norm_cfg),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img']),
        ])
]
```

(segformer_mit-b0_512x512)

**Comment by Junjun2016** [comment]
> > What's the best result on your custom dataset (SOTA)?
> 
> The best result is 0.5814 ,which is 512x512 img scale for Mosaic(prob=1) augmentation.
> 
> Here is config script for more detail.
> 
> ```
> train_pipeline = [
>     dict(type='RandomMosaic', prob=1, img_scale=(512,512)),
>     dict(type='RandomCrop', crop_size=(512, 512)),
>     dict(type='RandomFlip', prob=0.5),
>     dict(type='Normalize', **img_norm_cfg),
>     dict(type='DefaultFormatBundle'),
>     dict(type='Collect', keys=['img', 'gt_semantic_seg']),
> ]
> test_pipeline = [
>     dict(type='LoadImageFromFile'),
>     dict(
>         type='MultiScaleFlipAug',
>         img_scale=(512,512),
>         flip=False,
>         transforms=[
>             dict(type='Resize', keep_ratio=True),
>             dict(type='RandomFlip'),
>             dict(type='Normalize', **img_norm_cfg),
>             dict(type='ImageToTensor', keys=['img']),
>             dict(type='Collect', keys=['img']),
>         ])
> ]
> ```
> 
> (segformer_mit-b0_512x512)

I mean the best result you got before.
Can mosaic augmentation boost the best result?

**Comment by Younghoon-Lee** [comment]
> I mean the best result you got before. Can mosaic augmentation boost the best result?
Oh I see. 

Unfortunately, now we only have the best result which had been applied many tricks(due to competition),so it is hard to tell the best result of the model itself. 

We will do test right away with same condition as before .

**Comment by Younghoon-Lee** [comment]
Here's some results we have done so far.

|exp_num|method|mIoU best|best epoch|mIoU last|last epoch|
|:---|:---:|:---:|:---:|:---:|:---:|
|1(**SOTA**)|swinL_segformer_512x512_epoch18|0.7627|11|0.7611|18|
|2|swinL_segformer_512x512_epoch18_mosaic_prob_0.2|0.7588|11|0.7571|18|
|3|swinL_segformer_512x512_epoch18_mosaic_prob_0.3|0.7577|14|0.7558|18|
|4|swinL_segformer_512x512_epoch18_mosaic_prob_0.5|0.7567|13|0.7566|18|
|5|swinL_segformer_512x512_epoch18_mosaic_prob_1.0|0.755|18|0.755|18|

It seems that Mosaic couldn't boost our best result. 
As we keep doing our test to boost the best result, we will keep you update if anything comes out.

**Comment by Junjun2016** [comment]
> > I mean the best result you got before. Can mosaic augmentation boost the best result?
> 
> Oh I see.
> 
> Unfortunately, now we only have the best result which had been applied many tricks(due to competition),so it is hard to tell the best result of the model itself.
> 
> We will do test right away with same condition as before .

That may not be a good data augmentation strategy for segmentation.

**Comment by Younghoon-Lee** [comment]
> > > I mean the best result you got before. Can mosaic augmentation boost the best result?
> > 
> > 
> > Oh I see.
> > Unfortunately, now we only have the best result which had been applied many tricks(due to competition),so it is hard to tell the best result of the model itself.
> > We will do test right away with same condition as before .
> 
> That may not be a good data augmentation strategy for segmentation.

So we are just using SOTA model(without any tricks) we used and doing some augmentation tests only.

**Comment by Junjun2016** [comment]
> > > > I mean the best result you got before. Can mosaic augmentation boost the best result?
> > > 
> > > 
> > > Oh I see.
> > > Unfortunately, now we only have the best result which had been applied many tricks(due to competition),so it is hard to tell the best result of the model itself.
> > > We will do test right away with same condition as before .
> > 
> > 
> > That may not be a good data augmentation strategy for segmentation.
> 
> So we are just using SOTA model(without any tricks) we used and doing some augmentation tests only.

Make sense.

**Comment by Younghoon-Lee** [comment]
Here's some results we have done.

|exp_num|method|mIoU best|best epoch|mIoU last|last epoch|
|:---|:---:|:---:|:---:|:---:|:---:|
|1(**SOTA**)|512x512_epoch18|0.7627|11|0.7611|18|
|2|512x512_epoch18_mosaic_prob_0.2|0.7588|11|0.7571|18|
|3|512x512_epoch18_mosaic_prob_0.3|0.7577|14|0.7558|18|
|4|512x512_epoch18_mosaic_prob_0.5|0.7567|13|0.7566|18|
|5|512x512_epoch18_mosaic_prob_1.0|0.755|18|0.755|18|

Actually, we did several more tests such as without crop, changing center ratio, img scale etc .
But It seems that Mosaic couldn't boost our best result.

**Review by Junjun2016 — APPROVED** [review]

File: mmseg/datasets/dataset_wrappers.py [hunk 3]
```diff
@@ -188,3 +190,85 @@ def __getitem__(self, idx):
...
+@DATASETS.register_module()
+class MultiImageMixDataset:
+    """A wrapper of multiple images mixed dataset.
+
+    Suitable for training on multiple images mixed data augmentation like
+    mosaic and mixup. For the augmentation pipeline of mixed image data,
+    the `get_indexes` method needs to be provided to obtain the image
+    indexes, and you can set `skip_flags` to change the pipeline running
+    process.
+    Args:
```
Comment (MeowZheng) [inline mmseg/datasets/dataset_wrappers.py:204]:
```suggestion

    Args:
```

**Review by MeowZheng — APPROVED** [review]

File: mmseg/datasets/dataset_wrappers.py (hunk 3 above)
Comment (lkm2835) [reply mmseg/datasets/dataset_wrappers.py:204]: `pre-commit run --all-files`

```
flake8 .... Failed

line 204: W293 blank line contains whitespace
```

Is this okay?

Comment (MeowZheng) [reply mmseg/datasets/dataset_wrappers.py:204]: Just add a blank line, there might be some whitespace.
Please refer
https://github.com/open-mmlab/mmsegmentation/blob/ff9d695b4051e9f2384da11ba4dbdef5d003a856/mmseg/datasets/dataset_wrappers.py#L41-L54
